### ISSUE TITLE
Allow custom lazy scanning from python

### ISSUE THREAD

**Issue by OneRaynyDay** [pola-rs/polars#4351]
#### Describe your feature request

My understanding of the polars query planner is modest at best, so please correct me if I'm (ab)using the terms used below or if I have a conceptual misunderstanding:

`pl.scan_parquet` is great for creating a `pl.LazyFrame` from a large collection of parquet files and push down predicates/filters/projections/etc. Polars also support a fairly wide selection of reads/scans from other data sources. Is it possible to allow someone to pass in a python function that takes in predicates/filters/projections/etc and returns a `pl.LazyFrame`? We would consider it a part of a query plan and is essentially a backdoor to augmenting the query with extra optimizations, extra data source reads, extra asserts, etc.

If I'm not mistaken, [this particular PythonScanExec](https://github.com/pola-rs/polars/blob/1e3f703698099faff5cddd2d96e4cbd52944bb93/polars/polars-lazy/src/physical_plan/executors/python_scan.rs#L19) actually allows you to call an arbitrary function `PyBytes::new(py, &self.options.scan_fn)` and get the output. It seems like it only takes in the columns but not the predicates/filters/projections/etc, and I'm not sure if this interface is exposed in python. Would we be able to build on top of this?

**Comment by ritchie46** [comment]
Yes, you can use that node. This [file](https://github.com/pola-rs/polars/blob/master/py-polars/polars/internals/anonymous_scan.py) is an example. There we create a `scan_ds` and a `scan_parquet_fsspec` function.

We push projections to the node and pass that as a `with_columns: list[str]` argument. The predicates are pushed down to after that node. I am not really sure how we can pass a predicate to the function itself and what benefit does it have.

What I mean by that is that predicates internally are represented as a virtual function that produces a boolean mask. Not something that's usable on the python side. And if we pass the predicate as expression then you have to apply that predicate in the function itself and that does not seem to have much benefit to us applying that predicate when that function is finished.

**Comment by universalmind303** [comment]
This is already supported in Rust via `AnonymousScan` trait. This trait allows for predicate, projection & slice pushdowns. We could potentially look at exposing a python interface to interact with that trait?

**Comment by OneRaynyDay** [comment]
Yeah I think having a python interface would be awesome! Currently the effort to build rust at my company is nascent at best, so having a python fn drop-in would require the least amount of hoops to jump

**Comment by cmdlineluser** [comment]
I may be mistaken, but IO Plugins were recently added - do they solve this?

- https://github.com/pola-rs/polars/pull/17939

Relevant test that was added:

- https://github.com/pola-rs/polars/blob/main/py-polars/tests/unit/io/test_plugins.py

**Comment by tmct** [comment]
I didn't know about those! Will have to have a look, thanks

**Comment by tmct** [comment]
That does indeed look extremely promising, most notably that it's more of a public API. I'll use those. Thanks again

It looks like the "lazy schema" isn't ready yet - do you know if there is an issue to track that?

>    # TODO: make lazy via callable

**Comment by tmct** [comment]
And - sorry for my unfamiliarity - do you know what it means for the iterator to return multiple DataFrames? Is this for a scenario where scan_my_source returns multiple LazyFrames in a tuple?

**Comment by tmct** [comment]
Thanks to some help from @itamarst I now have a branch of my fork with this feature on it! It's not prod-ready at this point but I will have a play with it.

Example (passing) test from `poc` branch: https://github.com/tmct/polars/blob/8773d9c/py-polars/tests/unit/io/test_plugins.py#L39

**Comment by tmct** [comment]
Ah, I meant to post this in #18638, sorry, will post there

**Comment by coastalwhite** [comment]
I think this issue is closed by IO plugins. If there are any specific features you want to see, please open a new issue.
